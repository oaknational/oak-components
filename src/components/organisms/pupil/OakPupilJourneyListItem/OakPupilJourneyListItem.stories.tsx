import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakPupilJourneyListItem } from "./OakPupilJourneyListItem";

import { OakFlex } from "@/components/atoms";

const meta: Meta<typeof OakPupilJourneyListItem> = {
  component: OakPupilJourneyListItem,
  tags: ["autodocs"],
  title: "components/organisms/pupil/OakPupilJourneyListItem",
  args: {
    title: "Lesson 1",
    index: 1,
    href: "#",
  },
  decorators: [
    (Story) => {
      return (
        <OakFlex
          $flexDirection="column"
          $gap="space-between-m"
          $background={"bg-decorative4-main"}
          $pa={"inner-padding-xl"}
        >
          {Story()}
        </OakFlex>
      );
    },
  ],
  parameters: {
    controls: {
      include: [],
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakPupilJourneyListItem>;

export const Default: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} />,
};

export const Lessons: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} />,
  args: {
    numberOfLessons: 6,
  },
};

export const AsAButton: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} as="button" />,
};

export const Disabled: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} />,
  args: {
    disabled: true,
  },
};
