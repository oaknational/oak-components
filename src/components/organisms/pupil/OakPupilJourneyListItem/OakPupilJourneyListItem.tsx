import React, { ComponentPropsWithoutRef, ElementType } from "react";
import styled, { css } from "styled-components";

import { OakBox, OakFlex } from "@/components/atoms";
import { parseColor } from "@/styles/helpers/parseColor";
import { OakRoundIcon } from "@/components/molecules";
import { parseColorFilter } from "@/styles/helpers/parseColorFilter";
import { parseSpacing } from "@/styles/helpers/parseSpacing";
import { parseDropShadow } from "@/styles/helpers/parseDropShadow";

type OakPupilJourneyListItemProps<C extends ElementType> = {
  as?: C;
  /**
   * Disable the section preventing navigation to it.
   */
  disabled?: boolean;
  /**
   * shows that a section is unavailable
   */
  unavailable?: boolean;
  index: number;
  title: string;
  numberOfLessons?: number;
} & ComponentPropsWithoutRef<C>;

const StyledLabel = styled(OakBox)``;

const StyledRoundIcon = styled(OakRoundIcon)<{
  $disabled?: boolean;
}>`
  width: ${parseSpacing("all-spacing-8")};
  height: ${parseSpacing("all-spacing-8")};
  padding: 0;

  background: transparent;

  img {
    filter: ${(props) =>
      parseColorFilter(props.$disabled ? "icon-disabled" : "icon-inverted")};
  }
`;

const activeIconStyles = css`
  ${StyledRoundIcon} {
    box-shadow: ${parseDropShadow("drop-shadow-lemon")},
      ${parseDropShadow("drop-shadow-grey")};
  }
`;

const hoverIconStyles = css`
  ${StyledRoundIcon} {
    background: ${parseColor("bg-btn-primary")};

    img {
      filter: ${parseColorFilter("icon-main")};
    }
  }
`;

const StyledLessonNavItem = styled(OakFlex)<{ $disabled?: boolean }>`
  outline: none;
  text-align: initial;

  &:focus-visible {
    box-shadow: ${parseDropShadow("drop-shadow-centered-lemon")},
      ${parseDropShadow("drop-shadow-centered-grey")};
  }

  ${(props) => props.$disabled && "cursor: not-allowed"}

  ${(props) =>
    !props.$disabled &&
    css`
      cursor: pointer;

      /* Don't apply hover styles on touch devices */
      @media (hover: hover) {
        &:hover {
          background: ${parseColor("bg-decorative1-subdued")};
          ${StyledLabel} {
            text-decoration: underline;
          }

          ${hoverIconStyles}
        }
      }

      &:active {
        box-shadow: ${parseDropShadow("drop-shadow-lemon")},
          ${parseDropShadow("drop-shadow-grey")};
        ${activeIconStyles}
        ${hoverIconStyles}
      }
    `}
`;

const FlexedOakBox = styled(OakBox)`
  flex: 1;
`;

/**
 * Enables navigation to the given section of a lesson as well as displaying current progress
 */
export const OakPupilJourneyListItem = <C extends ElementType = "a">(
  props: OakPupilJourneyListItemProps<C>,
) => {
  const {
    as,
    lessonSectionName,
    progress,
    disabled,
    href,
    unavailable,
    onClick,
    ...rest
  } = props;

  const disabledOrUnavailable = disabled || unavailable;
  return (
    <StyledLessonNavItem
      as={disabledOrUnavailable ? "div" : as ?? "a"}
      $gap={["space-between-s", "space-between-m2"]}
      $alignItems="center"
      $justifyContent={"space-between"}
      $flexWrap={"wrap"}
      $background={unavailable ? "bg-neutral" : "bg-primary"}
      $pa={["inner-padding-l", "inner-padding-xl"]}
      $borderRadius="border-radius-m"
      $ba={unavailable ? "border-solid-m" : "border-solid-none"}
      $borderColor={unavailable ? "border-neutral-lighter" : "transparent"}
      $disabled={disabledOrUnavailable}
      $color="text-primary"
      href={disabledOrUnavailable ? undefined : href}
      onClick={disabledOrUnavailable ? undefined : onClick}
      {...rest}
    >
      <OakFlex $alignItems={"baseline"} $gap={["space-between-m2"]}>
        {" "}
        <OakFlex>
          <OakBox
            $font={["heading-5", "heading-4"]}
            $color={props.unavailable ? "text-disabled" : "text-primary"}
            $textDecoration={"none"}
          >
            {props.index}
          </OakBox>
        </OakFlex>
        <FlexedOakBox>
          <StyledLabel
            $font={["heading-6", "heading-5"]}
            $color={disabledOrUnavailable ? "text-disabled" : "text-primary"}
          >
            {props.title}
          </StyledLabel>
        </FlexedOakBox>
      </OakFlex>

      <OakFlex
        $alignItems={"center"}
        $gap={"space-between-xs"}
        $flexBasis={"auto"}
        $flexGrow={1}
        $justifyContent={"flex-end"}
      >
        {props.numberOfLessons && !props.unavailable && (
          <StyledLabel
            $font={"heading-7"}
            $color={disabledOrUnavailable ? "text-disabled" : "text-primary"}
          >
            {props.numberOfLessons} lessons
          </StyledLabel>
        )}
        {props.unavailable && (
          <StyledLabel
            $font={"heading-7"}
            $color={disabledOrUnavailable ? "text-disabled" : "text-primary"}
          >
            Unavailable
          </StyledLabel>
        )}
        {!props.unavailable && (
          <StyledRoundIcon
            iconName="chevron-right"
            $disabled={disabledOrUnavailable}
          />
        )}
      </OakFlex>
    </StyledLessonNavItem>
  );
};
