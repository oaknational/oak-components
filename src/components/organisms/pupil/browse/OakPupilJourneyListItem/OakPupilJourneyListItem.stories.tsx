import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakPupilJourneyListItem } from "./OakPupilJourneyListItem";

import { OakFlex } from "@/components/atoms";

const meta: Meta<typeof OakPupilJourneyListItem> = {
  component: OakPupilJourneyListItem,
  tags: ["autodocs"],
  args: {
    title: "Lesson 1",
    index: 1,
    href: "#",
  },
  argTypes: {
    title: { control: { type: "text" } },
    index: { control: { type: "number" } },
    numberOfLessons: { control: { type: "number" } },
    disabled: { control: { type: "boolean" } },
    unavailable: { control: { type: "boolean" } },
  },
  decorators: [
    (Story) => {
      return (
        <OakFlex
          $flexDirection="column"
          $gap="spacing-24"
          $background={"bg-decorative4-main"}
          $pa={"spacing-24"}
          role="list"
        >
          {Story()}
        </OakFlex>
      );
    },
  ],
  parameters: {
    controls: {
      include: ["title", "index", "numberOfLessons", "disabled", "unavailable"],
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakPupilJourneyListItem>;

export const Default: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} />,
};

export const Lessons: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} />,
  args: {
    numberOfLessons: 6,
  },
};

export const ReallyLongTitle: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} />,
  args: {
    title:
      "This is a really long title that should wrap around to the next line",
    numberOfLessons: 6,
  },
};

export const AsAButton: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} as="button" />,
  args: {
    href: undefined,
  },
};

export const Disabled: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} />,
  args: {
    disabled: true,
  },
};

export const Unavailable: Story = {
  render: (args) => <OakPupilJourneyListItem {...args} />,
  args: {
    disabled: true,
    unavailable: true,
    numberOfLessons: 14,
  },
};
