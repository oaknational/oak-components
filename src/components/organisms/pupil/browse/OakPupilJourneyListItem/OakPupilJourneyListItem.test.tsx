import React from "react";
import "@testing-library/jest-dom";

import { OakPupilJourneyListItem } from "./OakPupilJourneyListItem";

import renderWithTheme from "@/test-helpers/renderWithTheme";

describe(OakPupilJourneyListItem, () => {
  it("matches snapshot", () => {
    const { container } = renderWithTheme(
      <OakPupilJourneyListItem as="a" title="Lesson 1" index={1} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a div when the item is disabled", () => {
    const { getByTestId } = renderWithTheme(
      <>
        <OakPupilJourneyListItem
          data-testid="intro"
          as="a"
          title="Lesson 1"
          index={1}
          disabled
        />
      </>,
    );

    expect(getByTestId("intro").tagName).toBe("DIV");
  });
  it("renders a button when the item is not disabled", () => {
    const { getByTestId } = renderWithTheme(
      <>
        <OakPupilJourneyListItem
          data-testid="intro"
          as="a"
          title="Lesson 1"
          index={1}
        />
      </>,
    );

    expect(getByTestId("intro").tagName).toBe("A");
  });
  it("renders the number of lessons when provided", () => {
    const { getByTestId } = renderWithTheme(
      <>
        <OakPupilJourneyListItem
          data-testid="intro"
          as="a"
          title="Lesson 1"
          index={1}
          numberOfLessons={6}
        />
      </>,
    );

    expect(getByTestId("intro").textContent).toContain("6");
  });
});
