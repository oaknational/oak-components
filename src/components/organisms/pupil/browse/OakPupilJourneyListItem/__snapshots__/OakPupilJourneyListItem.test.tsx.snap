// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakPupilJourneyListItem matches snapshot 1`] = `
.c0 {
  font-family: --var(google-font),Lexend,sans-serif;
}

.c1 {
  padding: 1.25rem;
  color: #222222;
  background: #ffffff;
  border: 0rem solid;
  border-color: transparent;
  border-radius: 0.375rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c6 {
  color: #222222;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 600;
  font-size: 1.5rem;
  line-height: 2rem;
  -webkit-letter-spacing: 0.0115rem;
  -moz-letter-spacing: 0.0115rem;
  -ms-letter-spacing: 0.0115rem;
  letter-spacing: 0.0115rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8 {
  color: #222222;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 600;
  font-size: 1.25rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: 0.0115rem;
  -moz-letter-spacing: 0.0115rem;
  -ms-letter-spacing: 0.0115rem;
  letter-spacing: 0.0115rem;
}

.c10 {
  width: 3rem;
  height: 3rem;
  padding: 0.25rem;
  background: #ffe555;
  border-radius: 6.25rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c13 {
  position: relative;
  width: 100%;
  min-width: 100%;
  height: 100%;
  min-height: 100%;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c14 {
  object-fit: contain;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: 1rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 2rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  gap: 0.75rem;
}

.c12 {
  width: 2.5rem;
  height: 2.5rem;
  padding: 0;
  background: transparent;
}

.c12 img {
  -webkit-filter: invert(10%) sepia(1%) saturate(236%) hue-rotate(314deg) brightness(95%) contrast(91%);
  filter: invert(10%) sepia(1%) saturate(236%) hue-rotate(314deg) brightness(95%) contrast(91%);
}

.c3 {
  outline: none;
  text-align: initial;
  cursor: pointer;
}

.c3:focus-visible {
  box-shadow: 0 0 0 0.125rem rgba(255,229,85,100%), 0 0 0 0.3rem rgba(87,87,87,100%);
}

.c3:active {
  box-shadow: 0.125rem 0.125rem 0 rgba(255,229,85,100%), 0.25rem 0.25rem 0 rgba(87,87,87,100%);
}

.c3:active .c11 {
  box-shadow: 0.125rem 0.125rem 0 rgba(255,229,85,100%), 0.25rem 0.25rem 0 rgba(87,87,87,100%);
}

.c3:active .c11 {
  background: #222222;
}

.c3:active .c11 img {
  -webkit-filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
  filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

@media (min-width:750px) {
  .c1 {
    padding: 1.5rem;
  }
}

@media (min-width:750px) {
  .c6 {
    font-weight: 600;
  }
}

@media (min-width:750px) {
  .c6 {
    font-size: 2rem;
  }
}

@media (min-width:750px) {
  .c6 {
    line-height: 2.5rem;
  }
}

@media (min-width:750px) {
  .c6 {
    -webkit-letter-spacing: 0.0115rem;
    -moz-letter-spacing: 0.0115rem;
    -ms-letter-spacing: 0.0115rem;
    letter-spacing: 0.0115rem;
  }
}

@media (min-width:750px) {
  .c8 {
    font-weight: 600;
  }
}

@media (min-width:750px) {
  .c8 {
    font-size: 1.5rem;
  }
}

@media (min-width:750px) {
  .c8 {
    line-height: 2rem;
  }
}

@media (min-width:750px) {
  .c8 {
    -webkit-letter-spacing: 0.0115rem;
    -moz-letter-spacing: 0.0115rem;
    -ms-letter-spacing: 0.0115rem;
    letter-spacing: 0.0115rem;
  }
}

@media (min-width:750px) {
  .c2 {
    gap: 2rem;
  }
}

@media (hover:hover) {
  .c3:hover {
    background: #dff9de;
  }

  .c3:hover .c11 {
    background: #222222;
  }

  .c3:hover .c11 img {
    -webkit-filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
    filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
  }
}

<div>
  <div
    class="c0"
    role="listitem"
  >
    <a
      class="c1 c2 c3 pupil-journey-item"
      title="Lesson 1"
    >
      <div
        class="c0 c4"
      >
         
        <div
          class="c0 c5"
        >
          <div
            class="c6"
          >
            1
          </div>
        </div>
        <div
          class="c0 c7"
        >
          <div
            class="c8"
          >
            Lesson 1
          </div>
        </div>
      </div>
      <div
        class="c0 c9"
      >
        <div
          class="c10 c11 c12"
        >
          <div
            class="c13"
          >
            <img
              alt=""
              class="c14"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              src="https://res.cloudinary.com/mock-cloudinary-cloud/image/upload/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;
